### PR TITLE
Add tests for Transports.AspNetCore, increasing coverage to 99%

### PR DESCRIPTION
**Pull request by Shane32**
## Cleanup build props
- Apply NRT to entire solution
- Remove unneeded NoWarn entries
- Set all warnings as errors
- Internals visible to test project

## Tests
- Remove some old test projects
- Replace chat schema (since my tests heavily depend on it)
- Add new test project

## Coverage

| Module                               | Line   | Branch | Method |
|--------------------------------------|--------|--------|--------|
| GraphQL.Server.Samples.Schemas.Chat  | 91.25% | 90%    | 96.42% |
| GraphQL.Server.Transports.AspNetCore | 97.81% | 92.94% | 99.64% |
| GraphQL.Server.Ui.Altair             | 23.43% | 16.66% | 50%    |
| GraphQL.Server.Ui.GraphiQL           | 25%    | 14.28% | 55.55% |
| GraphQL.Server.Ui.Playground         | 28.57% | 14.7%  | 50%    |
| GraphQL.Server.Ui.Voyager            | 22.58% | 16.66% | 46.66% |

### CODE REVIEW

**Comment by Shane32** [comment]
@sungam3r As promised, here are the tests to accompany the code that I've submitted over the last month or two, with coverage around 98%.  It's a lot of code; over 7,500 lines.  Likely as not you will want to review every file.  If not, I would be happy to direct your attention to the files that I think you would be most interested in reviewing:

- All csproj files, props and targets files
- GitHub workflow yml file
- `Samples.Schema.Chat` files (you'd probably want to compare the new sample schema to the old sample schema)
- New `ExecutionResultActionResult.cs` file
- Api approvals
- `AuthorizationTests.cs` (the one in the Middleware folder)
- `AuthorizationTests.cs` (the other one, for the validation rule)

I have made heavy use of mocks to test the websocket implementation code.  Yet, there is full end-to-end tests in the following places, which I suggest you review:
- `GetTests.cs`
- `PostTests.cs`
- `BatchTests.cs`
- `ChatTests.cs` (which covers WebSocket implementations)

In addition to the mundane tests to cover the extension methods, websocket implementation details, and so forth, which mostly rely on mocking, I've also included CORS tests, to be sure the ASP.Net Core CORS implementation plays nicely with this library, and a FileUpload test, simulating how one can derive from `GraphQLHttpMiddleware` to allow a multipart form upload submission to include files that are injected as variables to the GraphQL engine (versus a base64 encoding or similar).

I did not write code to test the UI packages, but I left the old Samples.Server.Tests code project, which is exercising some of the UI packages' code.

File: tests/Transports.AspNetCore.Tests/ChatTests.cs
Comment (Shane32) [inline tests/Transports.AspNetCore.Tests/ChatTests.cs:170]: I would like to remove this delay, but the test then typically fails.  I believe this is because the prior call to `SendMessageAsync` gets processed more slowly than the following message via `AddMessageInternal`.  The latter call runs on a separate thread over a separate HttpClient, so there is no synchronization that ensures that the prior subscription request has completed before posting the new message.  Since the subscription server does not send any response, I cannot simply wait for a response here.  My only thought (which I have not tried) is that perhaps if some subsequent message that requires a response was sent via the WebSocket connection (e.g. CountMessages), then we could await that response before sending the add-message via a separate HttpClient.  Since there is no asynchronous code present in the chat service, this should eliminate the need for Task.Delay.  Still this is not a guaranteed fix because the BaseSubscriptionHandler is designed to process requests over the same WebSocket connection asynchronously. (See #821 )

Comment (Shane32) [reply tests/Transports.AspNetCore.Tests/ChatTests.cs:170]: This same issue (necessary call to Task.Delay) is present elsewhere in this file and also in `TestServerExtensions.cs`

File: samples/Samples.Schemas.Chat/Schema/ChatSchema.cs
Comment (Shane32) [inline samples/Samples.Schemas.Chat/Schema/ChatSchema.cs:3]: My tests do not use this file, but rather uses `AddAutoSchema` to configure the schema; but the old `Samples.Server.Tests` project does.  I can probably eliminate this file and update the `Samples.Server.Tests` project if you like.  This is the only "code-first" file here; everything else is "type-first", making the sample very easy to read.

**Comment by Shane32** [comment]
@sungam3r I have removed the added feature and moved it to a separate PR.  There is only one line of executable code added - a simple null check.  As this PR simply comprises of tests, I will merge this PR if you do not have time to review it.  Please let me know if you plan to review it soon.

**Comment by sungam3r** [comment]
I'll review all pending PRs from you in coming weekend.

File: samples/Samples.Schemas.Chat/Schema/ChatSchema.cs
Comment (sungam3r) [reply samples/Samples.Schemas.Chat/Schema/ChatSchema.cs:3]: > I can probably eliminate this file and update the Samples.Server.Tests project if you like.  ... making the sample very easy to read

50/50 here,

File: Directory.Build.targets
Comment (sungam3r) [inline Directory.Build.targets:7]: 👍

File: Directory.Build.targets
Comment (sungam3r) [inline Directory.Build.targets:10]: CS1998 ?

File: Directory.Build.props
Comment (sungam3r) [inline Directory.Build.props:25]:
```suggestion
    <NoWarn>$(NoWarn);IDE0056;IDE0057</NoWarn> <!-- Index/Range operators -->
```

File: samples/Samples.Schemas.Chat/Samples.Schemas.Chat.csproj
Comment (sungam3r) [inline samples/Samples.Schemas.Chat/Samples.Schemas.Chat.csproj:12]: I would move implicit usings into separate group.

File: samples/Samples.Schemas.Chat/Samples.Schemas.Chat.csproj
Comment (sungam3r) [inline samples/Samples.Schemas.Chat/Samples.Schemas.Chat.csproj:4]: TargetFramework**S** ?

File: .github/workflows/test.yml
Comment (sungam3r) [inline .github/workflows/test.yml:66]: изображение
Looks odd.

File: samples/Samples.Server/CustomErrorInfoProvider.cs
Comment (sungam3r) [inline samples/Samples.Server/CustomErrorInfoProvider.cs:29]: Why `!` operator?

File: samples/Samples.Server/StartupWithRouting.cs
Comment (sungam3r) [inline samples/Samples.Server/StartupWithRouting.cs:51]: Also check README.md for API changes and other stuff.

File: src/Transports.AspNetCore/Assembly.cs
Comment (sungam3r) [inline src/Transports.AspNetCore/Assembly.cs:3]: @Shane32 What I have learnt not so much time ago - https://www.meziantou.net/declaring-internalsvisibleto-in-the-csproj.htm  `InternalsVisibleTo` attribute became "native" since NET5.

File: tests/Samples.Server.Tests/Serializer.cs
Comment (sungam3r) [inline tests/Samples.Server.Tests/Serializer.cs:54]: ?

File: tests/Transports.AspNetCore.Tests/AuthorizationTests.cs
Comment (sungam3r) [inline tests/Transports.AspNetCore.Tests/AuthorizationTests.cs:635]:
```suggestion
    // these tests will fail core validation rules and so it does not really matter
```

File: samples/Samples.Server/CustomErrorInfoProvider.cs
Comment (Shane32) [reply samples/Samples.Server/CustomErrorInfoProvider.cs:29]: `error.PolicyAuthorizationResult` may be null if a role check failed rather than a policy check.  `Failure` is defined as nullable by .NET because `AuthorizationResult` may be a successful result -- but it will never be so here.

This code is old unchanged code.  I'll wrap the `switch` in an `if` to make a better sample.  `Failure!` is still proper.

File: tests/Transports.AspNetCore.Tests/ChatTests.cs
Comment (sungam3r) [reply tests/Transports.AspNetCore.Tests/ChatTests.cs:170]: Anyway it  would be better to place a comment

File: samples/Samples.Server/CustomErrorInfoProvider.cs
Comment (sungam3r) [reply samples/Samples.Server/CustomErrorInfoProvider.cs]: ok

File: tests/Samples.Server.Tests/Serializer.cs
Comment (Shane32) [reply tests/Samples.Server.Tests/Serializer.cs:54]: It's perhaps a little stupid, but due to nullability differences, it won't compile as it was before.  Dictionaries can't have nullable keys, so I can't redefine the key of the dictionary to `string?`.

File: .github/workflows/test.yml
Comment (Shane32) [reply .github/workflows/test.yml:66]: Yeah i guess.... I didn't change that.

File: src/Transports.AspNetCore/Assembly.cs
Comment (Shane32) [reply src/Transports.AspNetCore/Assembly.cs:3]: Are you suggesting we convert to csproj prop?

File: samples/Samples.Server/StartupWithRouting.cs
Comment (Shane32) [reply samples/Samples.Server/StartupWithRouting.cs:51]: I'm planning to rewrite the readme, subject to your review.

File: tests/Samples.Server.Tests/Serializer.cs
Comment (sungam3r) [reply tests/Samples.Server.Tests/Serializer.cs:54]: изображение
Seems an issue for old tfms.

File: src/Transports.AspNetCore/Assembly.cs
Comment (sungam3r) [reply src/Transports.AspNetCore/Assembly.cs:3]: yes

File: Directory.Build.targets
Comment (Shane32) [reply Directory.Build.targets:10]: CS1998 won't let you write async methods if they don't contain await.  Which although makes sense for the main project, often causes a decrease in readability of tests. 

For example, when a delegate is `Func<ContextClass, Task<object>>` you can't write `async context => context.Widget`, but rather are required to write `context => Task.FromResult<object>(context.Widget)`.

So for tests CS1998 is disabled.

File: tests/Transports.AspNetCore.Tests/ChatTests.cs
Comment (Shane32) [reply tests/Transports.AspNetCore.Tests/ChatTests.cs:170]: ok i'll add a comment along the lines of what's written above

File: tests/Transports.AspNetCore.Tests/Middleware/BatchTests.cs
Comment (sungam3r) [inline tests/Transports.AspNetCore.Tests/Middleware/BatchTests.cs:173]:
```suggestion
        using var response = await PostBatchRequestAsync(new GraphQLRequest { Query = "mutation{clearMessages}" });
```

and for other usages

File: tests/Samples.Server.Tests/Serializer.cs
Comment (Shane32) [reply tests/Samples.Server.Tests/Serializer.cs:54]: Probably.

File: tests/Transports.AspNetCore.Tests/Middleware/BatchTests.cs
Comment (Shane32) [reply tests/Transports.AspNetCore.Tests/Middleware/BatchTests.cs:173]: I agree and will accept suggestions where you make them along these lines.  I'm not searching the tests to change them.

File: tests/Transports.AspNetCore.Tests/ChatTests.cs
Comment (Shane32) [reply tests/Transports.AspNetCore.Tests/ChatTests.cs:170]:
```suggestion
        // Delay is necessary to allow time for the asynchronous websocket handler code
        // to execute prior to the independent call to AddMessageInternal below.
        // Since the websocket call does not return a response when the subscription
        // has completed being set up, there is no response we can await to determine
        // when to call AddMessageInternal.
        await Task.Delay(1000);
```

Comment (Shane32) [reply tests/Transports.AspNetCore.Tests/ChatTests.cs:170]: is this sufficient?

Comment (sungam3r) [reply tests/Transports.AspNetCore.Tests/ChatTests.cs:170]: yes

File: src/Transports.AspNetCore/Assembly.cs
Comment (Shane32) [reply src/Transports.AspNetCore/Assembly.cs]: Changed.

File: samples/Samples.Schemas.Chat/Schema/ChatSchema.cs
Comment (Shane32) [reply samples/Samples.Schemas.Chat/Schema/ChatSchema.cs:3]: I'll leave it as-is.

File: tests/Transports.AspNetCore.Tests/Middleware/MiscTests.cs
Comment (sungam3r) [inline tests/Transports.AspNetCore.Tests/Middleware/MiscTests.cs:33]:
```suggestion
```

File: tests/Transports.AspNetCore.Tests/TestServerExtensions.cs
Comment (sungam3r) [inline tests/Transports.AspNetCore.Tests/TestServerExtensions.cs:18]:
```suggestion
        var content = new StringContent(data, Encoding.UTF8, "application/graphql+json");
```
?

File: tests/Transports.AspNetCore.Tests/Transports.AspNetCore.Tests.csproj
Comment (sungam3r) [inline tests/Transports.AspNetCore.Tests/Transports.AspNetCore.Tests.csproj:4]: Let's solve this task by [Microsoft.NETFramework.ReferenceAssemblies](https://www.nuget.org/packages/Microsoft.NETFramework.ReferenceAssemblies.net461/) 

https://andrewlock.net/using-reference-assemblies-to-build-net-framework-libararies-on-linux-without-mono/

**Review by sungam3r — APPROVED** [review]
Tooooooo much for me for single read. At first I tried to check everything line by line, but very quickly I realized that it would take up a lot of time. I believe that you did everything to write good tests. 👍

File: tests/Transports.AspNetCore.Tests/Transports.AspNetCore.Tests.csproj
Comment (Shane32) [reply tests/Transports.AspNetCore.Tests/Transports.AspNetCore.Tests.csproj:4]: Sounds good; I'll add an issue so I can investigate and test

Comment (Shane32) [reply tests/Transports.AspNetCore.Tests/Transports.AspNetCore.Tests.csproj:4]: #845

**Pull request merged**
